// vector2/vector2.js
var PI_2 = Math.PI / 2;

class Vector2 {
	constructor(x, y){
		this.x = x || 0;
		this.y = y || 0;
	}
	
	static get zero(){
		return new Vector2();
	}
	get isZero(){
		return this.equals(Vector2.zero);
	}
	
	static get up(){
		return new Vector2(0,1);
	}
	static get down(){
		return new Vector2(0,-1);
	}
	static get left(){
		return new Vector2(-1,0);
	}
	static get right(){
		return new Vector2(1,0);
	}
	
	//---------------------------------- ROUND
	
	static round(n, depth = 1000){
		return Math.round(depth * n) / depth;
	}
	static roundVector(v, depth = 1000){
		return v.clone().roundSelf(depth);
	}
	
	roundSelf(depth){
		return this.withSelf( (coo) => Vector2.round(coo, depth) );
	}
	
	//---------------------------------- OPERATIONS
	
	withSelf(forCoo = ()=>{}){
		this.x = forCoo(this.x);
		this.y = forCoo(this.y);
		return this;
	}
	
	clone(){
		return new Vector2(this.x, this.y);
	}
	
	static check(v){
		if (!v) return Vector2.zero;
		if (v instanceof Array) return new Vector2(v[0], v[1]);
		if (typeof v != 'object') return new Vector2(v, v);
		return v;
	}
	
	add(v){
		v = Vector2.check(v);
		this.x += v.x;
		this.y += v.y;
		return this;
	}
	sub(v){
		v = Vector2.check(v);
		this.x -= v.x;
		this.y -= v.y;
		return this;
	}
	mult(v){
		v = Vector2.check(v);
		this.x *= v.x;
		this.y *= v.y;
		return this;
	}
	normalize(){
		this.mult(1.0 / this.length);
		return this;
	}
	
	//---------------------------------- OTHER
	
	static isNumInRange(n, min, max){
		return n > min && n < max;
	}
	
	inRange(v, radius){
		return Vector2.isNumInRange(this.x, v.x - radius, v.x + radius) &&
			Vector2.isNumInRange(this.y, v.y - radius, v.y + radius);
	}
	
	equals(v, depth){
		if (depth !== undefined){
			return Vector2.roundVector(this, depth)
				.equals( Vector2.roundVector(v, depth) );
		}
		return this.x == v.x && this.y == v.y;
	}
	
	get sqrLength(){
		return this.x * this.x + this.y * this.y;
	}
	get length(){
		return Math.sqrt(this.sqrLength);
	}
	
	toString(){
		return '(' + this.x + ', ' + this.y + ')';
	}
	
	getAngle(){
		let angle = Math.acos(this.x / this.length);
		if (this.y < 0) angle = -angle;
		return angle;
	}
	
	setAngle(newAngle, center){
		if (!center) center = Vector2.zero;
		
		this.sub(center);
		let len = this.length;
		
		this.x = len * Math.cos(newAngle);
		this.y = len * Math.sin(newAngle);
		
		this.add(center);
		return this;
	}
}

export default Vector2;
