// canvas/vector2.js
class Vector2 {
	constructor(x, y){
		this.x = x || 0;
		this.y = y || 0;
	}
	
	static get zero(){
		return new Vector2();
	}
	static get up(){
		return new Vector2(0,1);
	}
	static get down(){
		return new Vector2(0,-1);
	}
	static get left(){
		return new Vector2(-1,0);
	}
	static get right(){
		return new Vector2(1,0);
	}
	
	clone(){
		return new Vector2(this.x, this.y);
	}
	
	check(v){
		if (!v) return Vector2.zero;
		if (v instanceof Array) return new Vector2(v[0], v[1]);
		if (typeof v != 'object') return new Vector2(v, v);
		return v;
	}
	
	add(v){
		v = this.check(v);
		this.x += v.x;
		this.y += v.y;
		return this;
	}
	sub(v){
		v = this.check(v);
		this.x -= v.x;
		this.y -= v.y;
		return this;
	}
	mult(v){
		v = this.check(v);
		this.x *= v.x;
		this.y *= v.y;
		return this;
	}
	normalize(){
		this.mult(1.0 / this.length);
		return this;
	}
	
	equals(v){
		return this.x == v.x && this.y == v.y;
	}
	
	get sqrLength(){
		return this.x * this.x + this.y * this.y;
	}
	get length(){
		return Math.sqrt(this.sqrLength);
	}
	
	toString(){
		return '(' + this.x + ', ' + this.y + ')';
	}
	
	rotate(newAngle, center){
		if (!center) center = Vector2.zero;
		
		this.sub(center);
		let len = this.length;
		
		this.x = len * Math.cos( newAngle );
		this.y = len * Math.sin( newAngle );
		
		this.add(center);
		return this;
	}
}

export default Vector2;
